packages/browser/src/Session.spec.ts

```diff
@@ -922,6 +940,70 @@ describe("Session", () => {
       expect(window.localStorage.getItem(KEY_CURRENT_URL)).toBeNull();
       expect(clientAuthentication.login).not.toHaveBeenCalled();
     });
+
+    it("does not retry silent authentication if an authentication error happens", async () => {
+      const sessionId = "mySession";
+      // Pretend that we have a previously active session.
+      mockLocalStorage({
+        [KEY_CURRENT_SESSION]: sessionId,
+      });
+      mockLocation("https://mock.current/location");
+      const mockedStorage = new StorageUtility(
+        mockStorage({
+          [`${USER_SESSION_PREFIX}:${sessionId}`]: {
+            isLoggedIn: "true",
+          },
+        }),
+        mockStorage({})
+      );
+      const clientAuthentication = mockClientAuthentication({
+        sessionInfoManager: mockSessionInfoManager(mockedStorage),
+      });
+      clientAuthentication.handleIncomingRedirect = jest
+        .fn(clientAuthentication.handleIncomingRedirect)
+        .mockImplementationOnce(async (_url, emitter) => {
+          emitter.emit(EVENTS.ERROR, "Some authentication error");
+          return {
+            isLoggedIn: false,
+            sessionId,
+          };
+        });
+      clientAuthentication.login = jest.fn();
+
+      const mySession = new Session({ clientAuthentication });
+      // eslint-disable-next-line no-void
+      void mySession.handleIncomingRedirect("https://some.redirect/url");
+      expect(mySession.info.isLoggedIn).toBe(false);
+      // The local storage should have been cleared by the auth error
+      expect(window.localStorage.getItem(KEY_CURRENT_SESSION)).toBeNull();
+      // Silent authentication should not have been attempted
+      expect(clientAuthentication.login).not.toHaveBeenCalled();
+    });
+
+    it("clears the session ID from local storage if the session expired", async () => {
+      const sessionId = "mySession";
+      // Pretend that we have a previously active session.
+      mockLocalStorage({
+        [KEY_CURRENT_SESSION]: sessionId,
+      });
+      mockLocation("https://mock.current/location");
+      const mockedStorage = new StorageUtility(
+        mockStorage({
+          [`${USER_SESSION_PREFIX}:${sessionId}`]: {
+            isLoggedIn: "true",
+          },
+        }),
+        mockStorage({})
+      );
+      const clientAuthentication = mockClientAuthentication({
+        sessionInfoManager: mockSessionInfoManager(mockedStorage),
+      });
+
+      const mySession = new Session({ clientAuthentication });
+      mySession.emit(EVENTS.SESSION_EXPIRED);
```

Comment:
Can we test this by actually inducing the expiry in mySession, instead of injecting the event? Just thinking if the code that produces this event may emit other events or change state in another way?

Reply:
I think it would be possible to mock this by actually expiring the session, by mocking a token response with zero time to live, but that would be more involved. Also, the portion of the code emitting this signal is meant to be isolated: it's the closure where the tokens/keys are kept. The signal is meant to be the only way for this closure to change state. Other signals could be emitted, but they should not impact the session state.